### PR TITLE
Implement calldata verification facet v1.0.0

### PR DESCRIPTION
**Pull request by ezynda3**

### CODE REVIEW

File: src/Facets/CalldataVerificationFacet.sol [hunk 1]
```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { ILiFi } from "../Interfaces/ILiFi.sol";
+import { LibSwap } from "../Libraries/LibSwap.sol";
+import { AmarokFacet } from "./AmarokFacet.sol";
+import { StargateFacet } from "./StargateFacet.sol";
+import { CelerIMFacet } from "./CelerIMFacet.sol";
+
+/// @title A title that should describe the contract/interface
```
Comment (satoshi-lifi) [inline src/Facets/CalldataVerificationFacet.sol:10]: Let's add actual title ;)

File: src/Facets/CalldataVerificationFacet.sol [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        ILiFi.BridgeData memory bridgeData;
+        (bridgeData) = abi.decode(data[4:], (ILiFi.BridgeData));
+
+        if (bridgeData.hasSourceSwaps) {
+            LibSwap.SwapData[] memory swapData;
+            (, swapData) = abi.decode(
+                data[4:],
+                (ILiFi.BridgeData, LibSwap.SwapData[])
+            );
+            sendingAssetId = swapData[0].sendingAssetId;
```
Comment (satoshi-lifi) [inline src/Facets/CalldataVerificationFacet.sol:96]: I think we always return the `bridgeData.sendingAssetId` for the `sendingAssetId`.
When we do swaps before the bridging, the final asset of swaps is `sendingAssetId`.

File: src/Facets/CalldataVerificationFacet.sol [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+pragma solidity 0.8.17;
+
+import { ILiFi } from "../Interfaces/ILiFi.sol";
+import { LibSwap } from "../Libraries/LibSwap.sol";
+import { AmarokFacet } from "./AmarokFacet.sol";
+import { StargateFacet } from "./StargateFacet.sol";
+import { CelerIMFacet } from "./CelerIMFacet.sol";
+
+/// @title A title that should describe the contract/interface
+/// @author Li.Finance (https://li.finance)
```
Comment (maxklenk) [inline src/Facets/CalldataVerificationFacet.sol:11]: /// @author LI.FI (https://li.fi)

File: src/Facets/CalldataVerificationFacet.sol (hunk 2 above)
Comment (maxklenk) [reply src/Facets/CalldataVerificationFacet.sol:96]: Yes, but we want to return which asset the user starts with in this case.

File: src/Facets/CalldataVerificationFacet.sol [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    function extractMainParameters(
+        bytes calldata data
+    )
+        public
+        pure
+        returns (
+            string memory bridge,
+            address sendingAssetId,
+            address receiver,
+            uint256 minAmount,
```
Comment (maxklenk) [inline src/Facets/CalldataVerificationFacet.sol:81]: I think we can name this `amount` as exactly this amount has to be passed.

File: src/Facets/CalldataVerificationFacet.sol [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                keccak256(abi.encodePacked(bridge))) &&
+            (sendingAssetId == 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF ||
+                bridgeData.sendingAssetId == sendingAssetId) &&
+            (receiver == 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF ||
+                bridgeData.receiver == receiver) &&
+            (amount == type(uint256).max || bridgeData.minAmount == amount) &&
+            (destinationChainId == type(uint256).max ||
+                bridgeData.destinationChainId == destinationChainId) &&
+            bridgeData.hasSourceSwaps == hasSourceSwaps &&
+            bridgeData.hasDestinationCall == hasDestinationCall;
```
Comment (maxklenk) [inline src/Facets/CalldataVerificationFacet.sol:160]: Splitting this in more lines with comments would help to understand what we check here.

File: src/Facets/CalldataVerificationFacet.sol [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                    LibSwap.SwapData[],
+                    CelerIMFacet.CelerIMData
+                )
+            );
+            return keccak256(dstCalldata) == keccak256(celerIMData.callData);
+        }
+
+        // All other cases
+        return false;
+    }
```
Comment (maxklenk) [inline src/Facets/CalldataVerificationFacet.sol:239]: We should add a function to parse calls to the `GenericSwapFacet` facet.

**Review by maxklenk — CHANGES_REQUESTED** [review]

File: src/Facets/CalldataVerificationFacet.sol (hunk 6 above)
Comment (ezynda3) [reply src/Facets/CalldataVerificationFacet.sol:246]: @maxklenk What would we like to return from this?
```
reciever
amount
sendingAssetId
```

Comment (maxklenk) [reply src/Facets/CalldataVerificationFacet.sol:246]: @ezynda3 I think input and output are important
```
sendingAssetId <= first swap asset
sendingAmount <= first swap amount
receiver 
receivingAssetId <= last swap asset
receivingMinAmount <= passed minAmount
```

**Pull request merged**